api4/user_test.go

```diff
@@ -2068,8 +2069,19 @@ func TestUserLoginMFAFlow(t *testing.T) {
 	})
 
 	t.Run("WithCorrectMFA", func(t *testing.T) {
-		t.Skip("Skipping test that fails randomly.")
-		code := dgoogauth.ComputeCode(secret.Secret, time.Now().UTC().Unix()/30)
+		secret3, err := th.App.GenerateMfaSecret(th.BasicUser.Id)
+		assert.Nil(t, err)
+
+		// Fake user has MFA enabled
+		if result := <-th.Server.Store.User().UpdateMfaActive(th.BasicUser.Id, true); result.Err != nil {
+			t.Fatal(result.Err)
+		}
+
+		if result := <-th.Server.Store.User().UpdateMfaSecret(th.BasicUser.Id, secret3.Secret); result.Err != nil {
+			t.Fatal(result.Err)
+		}
+
+		code := dgoogauth.ComputeCode(secret3.Secret, time.Now().UTC().Unix()/30)
```

Comment:
Fundamentally, I'm guessing that the race seems to arise from the fact that `time.Now()` changes between here and the code downstream from `th.Client.LoginWithMFA` below. Is our implementation such that we don't allow for overlap of codes? If so, I imagine this impacts real users from time to time as well. If we accepted prev, current, or next codes this problem would go away entirely, but I don't know if that's best practice or not.

Reply:
We currently support a window size of 3 (https://github.com/mattermost/mattermost-server/blob/master/services/mfa/mfa.go#L87) which should cover that, I am really unsure where the race is coming from currently. I suppose having multiple MFA providers where we could mock the implementation would be beneficial (similar to supporting Yubikey etc), but that would be a bigger refactor. I'll see when I can find for this again to retest, thanks for taking the time to test it Jesse!